### PR TITLE
Removing features with invalid geometries before printing

### PR DESCRIPTION
**Pull request by procrastinatio**
Trying to fix https://github.com/geoadmin/mf-geoadmin3/issues/3714

The application is sending invalid GeoJSON geometries to the print server. For the record, the [GeoJSON format](https://tools.ietf.org/html/rfc7946)

The following [KML](https://www.procrastinatio.org/kml/2points-multipolygons.kml) has invalid (multi)polygons

On prod, this KML makes print crash:
https://s.geo.admin.ch/71bfa62ee8

With this PR, invalid geometries are removed:

https://mf-geoadmin3.int.bgdi.ch/mom_geom_valid

https://mf-chsdi3.int.bgdi.ch/shorten/71bfa870ae

### CODE REVIEW

**Comment by loicgasser** [comment]
Nice in general for these kind of things a service would be nice, so that you can test, re-use it.

also for inspiration:
https://github.com/bjornharrtell/jsts/blob/c742f471bf3357cc42a8f7d2bb229ed71f7df03a/src/org/locationtech/jts/operation/polygonize/EdgeRing.js#L50

**Comment by loicgasser** [comment]
https://github.com/geoadmin/mf-geoadmin3/blob/master/src/components/GeomUtilsService.js#L47

also

**Comment by gjn** [comment]
This is good for me. Thanks alot!

**Comment by gjn** [comment]
Could you put it in service as @loicgasser suggested?

**Comment by procrastinatio** [comment]
OL3 has no problem displaying invalid polygon (not closed, only three coordinates, or with a null area), which produces invalid GeoJSON. All tests are failing.

But in the first place, the KML are bogus, they should be tested first.

**Comment by loicgasser** [comment]
yes, we should correct the kml before adding it to the map I agree.

**Comment by loicgasser** [comment]
but instead of removing them you should fix them.

File: src/components/GeomUtilsService.js [hunk 1]
```diff
@@ -102,7 +111,8 @@ goog.provide('ga_geomutils_service');
                    geometry.getLinearRings());
             } else if (geometry instanceof ol.geom.LinearRing ||
                 geometry instanceof ol.geom.LineString) {
-              isValid = !this.hasUniqueCoord(geometry.getCoordinates());
+              isValid = isValidLinearRing(geometry) &&
```
Comment (oterral) [inline src/components/GeomUtilsService.js]: geometry can be a LineString and a Line String with 3 coordinates is valid.

**Comment by gjn** [comment]
So this won't fix the original issue anymore?

(I know that the demand to fix this bug with removing the bad stuff is a hack and proper kml validation and handling should be implemented, but we can't do that now).

**Comment by procrastinatio** [comment]
The only goal is not to make the print server crash. If a polygon has less than 4 points, it is discard.

**Comment by gjn** [comment]
Print doesn't crash, but kml is also not shown as before:

https://mf-geoadmin3.int.bgdi.ch/mom_geom_valid/index.html?topic=ech&lang=de&bgLayer=ch.swisstopo.pixelkarte-grau&layers=KML%7C%7Chttps:%2F%2Fpublic.geo.admin.ch%2FyK1QZOcrSeiPFdjmnZJqhg&X=213797.61&Y=691638.32&zoom=7

https://map.geo.admin.ch/?X=214089.51&Y=692860.69&zoom=6&topic=ech&lang=de&bgLayer=ch.swisstopo.pixelkarte-farbe&layers=KML%7C%7Chttps:%2F%2Fpublic.geo.admin.ch%2FyK1QZOcrSeiPFdjmnZJqhg

@davidoesch How tolerant should we be?

**Comment by procrastinatio** [comment]
At least it is coherent between view and print. Otherwise, people will start to complain about features not being printed, and I will have to solve these case.

**Comment by gjn** [comment]
I'm convinced. Merge if @davidoesch agrees.

Effect of this PR: We stop showing invalid geometries that can't be printed.

**Comment by davidoesch** [comment]
MUST:
- Print and View: WYSIWYG 
- What has been drawn with our tools has to be printable

if those two are given:
+1 for merge

**Comment by gjn** [comment]
Thanks alot @procrastinatio !

**Pull request merged**